modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java

```diff
@@ -87,38 +91,33 @@ public void onResponse(ReindexTaskState newTaskState) {
-                                if (assignmentAttempts < MAX_ASSIGNMENT_ATTEMPTS) {
-                                    int nextAttempt = assignmentAttempts + 1;
-                                    logger.debug("Attempting to retry reindex task assignment write. Attempt number " + nextAttempt);
-                                    assign(listener);
-                                } else {
-                                    String message = "Failed to write allocation id to reindex task doc after " + MAX_ASSIGNMENT_ATTEMPTS
-                                        + "retry attempts";
-                                    logger.info(message, ex);
-                                    listener.onFailure(new ElasticsearchException(message, ex));
-                                }
+                                logger.debug("Failed to write to {} index on ASSIGNMENT due to version conflict, retrying now",
+                                    REINDEX_INDEX, ex);
+                                assign(listener, delay);
                             } else {
-                                String message = "Failed to write allocation id to reindex task doc";
-                                logger.info(message, ex);
-                                listener.onFailure(new ElasticsearchException(message, ex));
+                                TimeValue nextDelay = getNextDelay(delay);
+                                logger.info(new ParameterizedMessage("Failed to write to {} index on ASSIGNMENT, retrying in {}",
+                                        REINDEX_INDEX, nextDelay), ex);
+                                threadPool.schedule(() -> assign(listener, nextDelay), nextDelay, ThreadPool.Names.SAME);
                             }
                         }
                     });
                 } else {
-                    ElasticsearchException ex = new ElasticsearchException("A newer task has already been allocated");
-                    listener.onFailure(ex);
+                    logger.info(new ParameterizedMessage("Failed to write ASSIGNMENT due to newer allocation, will not retry"));
+                    listener.onFailure(new ElasticsearchException("A newer task has already been allocated"));
                 }
             }
 
             @Override
             public void onFailure(Exception ex) {
-                String message = "Failed to fetch reindex task doc";
-                logger.info(message, ex);
-                listener.onFailure(new ElasticsearchException(message, ex));
+                TimeValue nextDelay = getNextDelay(delay);
+                logger.info(new ParameterizedMessage("Failed to read from {} index on ASSIGNMENT, retrying in {}", REINDEX_INDEX,
+                        nextDelay), ex);
+                threadPool.schedule(() -> assign(listener, nextDelay), nextDelay, ThreadPool.Names.SAME);
```

Comment:
Not sure if this is where we end if the document no longer exists, but I think we will need to stop retrying in case the doc is gone. OK to do this in a follow-up.